### PR TITLE
Iterators (WIP)

### PR DESCRIPTION
**Pull request by benbjohnson**
## Overview

This pull request adds a stream-based approach to the query engine via iterators. This initial example only includes a `FloatIterator` and an implementation of `MinIterator` for floats. It shows how an aggregate iterator would be built and shows how iterators can be composed.

### Performance

The following is a benchmark to calculate a windowed `min()` across a variety of intervals. This uses 10 input iterators (which would be shards within InfluxDB).

```
$ goo test -v -bench=Iterator ./tsdb
PASS
BenchmarkFloatMinIterator_1m-4 	50000000	        34.3 ns/op	       0 B/op	       0 allocs/op
BenchmarkFloatMinIterator_10m-4	50000000	        34.1 ns/op	       0 B/op	       0 allocs/op
BenchmarkFloatMinIterator_1h-4 	50000000	        33.9 ns/op	       0 B/op	       0 allocs/op
BenchmarkFloatMinIterator_1d-4 	50000000	        33.8 ns/op	       0 B/op	       0 allocs/op
ok  	github.com/influxdb/influxdb/tsdb	22.762s
```

This is showing that it can run a `min()` across 1 minute, 10 minute, 1 hour, & 1 day intervals with about the same performance per point. These benchmarks ramp up to 50M points and operate at ~30M pts/sec.

### CODE REVIEW

File: tsdb/iterator.go [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+	// Consume all points from all iterators within the window.
+	var min *FloatValue
+	for _, input := range itr.inputs {
+		for {
+			v := input.Next()
+			if v == nil {
+				break
+			} else if !v.Time.Before(endTime) {
+				input.unread(v)
```
Comment (otoolep) [inline tsdb/iterator.go]: Nice, the lookahead pattern appears in this code. It's always so easy to follow.

**Comment by otoolep** [comment]
This looks pretty nice.

File: tsdb/iterator.go (hunk 1 above)
Comment (benbjohnson) [inline tsdb/iterator.go]: Thanks. There should really be a `NextBefore(t time.Time)` on the buffered iterators. It would help clean up the min iterator.

**Comment by benbjohnson** [comment]
Thanks, Philip. I also wanted to note on this PR that the windowed iterators work because they expect input iterators to send their windows in order (although points within those windows do not need to be in order).

**Comment by nathanielc** [comment]
How would it look if the windowing was performed outside of the FloatMinIterator? For my use case the data is already windowed and so the FloatMinIterator would just need to exhaust the iterator looking for the min.

Maybe create WindowedFloatIterator that does all the windowing separate from the MinIterator?

**Comment by benbjohnson** [comment]
@nathanielc If you're only looking for a single value then you can use the `MinIterator` with a `0` `Interval`. I'm not sure how I would be able to separate out windowing from the `MinIterator` since the iterator needs to know what the next window is in order to compute a min value for it.

**Comment by nathanielc** [comment]
Could you do something like this?

```
// Get iterator for underlying data
fitr := FloatIterator(...)
// Nest iterator in windowed iterator.
witr := WindowedIterator(fitr)
witr.Interval = time.Minute
// pass windowed iterator to MinIterator
minItr := NewFloatMinIterator(witr)
```

The implementation for WindowedItrator just returns nil once the window has been exhausted. And MinIterator reads all values from its iterator and returns the min.

I am probably missing something....

**Comment by benbjohnson** [comment]
Right now `nil` returns signify an EOF on the iterator. It would have to use some kind of marker if it wanted to distinguish between end-of-window and end-of-stream. I'm not sure that would be cleaner than how it is now. Can you show a pseudo-impl?

**Comment by nathanielc** [comment]
I was missing the fact that the MinIterator etc are reused for multiple windows, but it makes sense why.

Might be possible to just delimit windows with `nil` and multiple `nil`s in a row indicate EOF. Just a wild guess/suggestion. I need to think about this a bit more to see how it would fit with what I need.

Overall though I like the concept and it looks great.

**Comment by benbjohnson** [comment]
I added `tsdb.Join()` in 8c5c8b1234ab380f98bc5463eeb52be245186d92. This combines iterators so they all emit values at the same timestamps for every unique input timestamp.

This is useful for joining series together like:

```
SELECT sum(x.value) + sum(y.value) FROM x, y GROUP BY time(10m)
```

The join allows us to ensure that every `x.value` point and `y.value` point  has a matching point on the other side with the same timestamp (even if that other point's value is `null`). 

This is also one example of a join. Another type of join might be one that merges all values from one side with a range on the other side, such as in @corylanou's example of `SELECT`-clause subqueries:

```
SELECT value + (SELECT avg(value) FROM y GROUP BY time(10m)) FROM x
```

**Comment by benbjohnson** [comment]
Also, I added benchmarks for the join but it's harder to give representative numbers with a synthetic benchmark here. If there's no jitter between timestamps (as is in the case of group by intervals) then there's no allocations. However, if a point doesn't exist for an iterator for a given timestamp then, obviously, an allocation would have to occur to create that point.

So, best case scenario, the join can combine 10 separate iterators with no jitter in about `150ns` on my 2012 MBP. If every iterator has a different point then that's 9 allocations per point and it jumps to about `2µs`.

**Comment by benbjohnson** [comment]
I'm closing this one in order to do incremental PRs against the `query-engine` branch.

**Pull request closed without merging**